api/src/main/java/org/apache/iceberg/metrics/DefaultTimer.java

```diff
@@ -0,0 +1,154 @@
+
+/**
+ * A default {@link Timer} implementation that uses a {@link Stopwatch} instance internally to measure time.
+ */
+public class DefaultTimer implements Timer {
+  private final TimeUnit defaultTimeUnit;
+  private final AtomicLong count = new AtomicLong();
+  private Duration duration = Duration.ZERO;
+  private final AtomicReference<Stopwatch> stopwatchRef = new AtomicReference<>();
+
+  public DefaultTimer(TimeUnit timeUnit) {
+    Preconditions.checkArgument(null != timeUnit, "TimeUnit must be non-null");
+    this.defaultTimeUnit = timeUnit;
+  }
+
+  @Override
+  public long count() {
+    return count.get();
+  }
+
+  @Override
+  public Duration totalDuration() {
+    return duration;
+  }
+
+  @Override
+  public Sample start() {
+    boolean updated = stopwatchRef.compareAndSet(null, Stopwatch.createStarted());
+    Preconditions.checkState(updated, "startTimer() called multiple times");
+    return new DefaultSample(stopwatchRef, defaultTimeUnit);
+  }
+
+  @Override
+  public void record(long amount, TimeUnit unit) {
+    if (amount >= 0) {
+      this.duration = this.duration.plus(amount, toChronoUnit(unit));
+      count.incrementAndGet();
+    }
+  }
+
+  @Override
+  public <T> T record(Supplier<T> supplier) {
+    Sample sample = start();
+    try {
+      return supplier.get();
+    } finally {
+      sample.stop(this);
+    }
+  }
+
+  @Override
+  public <T> T recordCallable(Callable<T> callable) throws Exception {
+    Sample sample = start();
+    try {
+      return callable.call();
+    } finally {
+      sample.stop(this);
+    }
+  }
+
+  @Override
+  public void record(Runnable runnable) {
+    Sample sample = start();
+    try {
+      runnable.run();
+    } finally {
+      sample.stop(this);
+    }
+  }
+
+  @Override
+  public String toString() {
+    return MoreObjects.toStringHelper(this)
+        .add("defaultTimeUnit", defaultTimeUnit)
+        .add("count", count)
+        .add("duration", duration)
+        .add("current", stopwatchRef)
+        .toString();
+  }
+
+  private static class DefaultSample implements Sample {
+    private final AtomicReference<Stopwatch> stopwatchRef;
+    private final TimeUnit defaultTimeUnit;
+
+    private DefaultSample(AtomicReference<Stopwatch> stopwatchRef, TimeUnit defaultTimeUnit) {
+      this.stopwatchRef = stopwatchRef;
+      this.defaultTimeUnit = defaultTimeUnit;
+    }
+
+    public void stop(Timer timer) {
+      Preconditions.checkArgument(null != timer, "Timer must be non-null");
+      Stopwatch stopwatch = stopwatchRef.getAndSet(null);
+      Preconditions.checkState(null != stopwatch, "startTimer() was not called");
+      timer.record(stopwatch.stop().elapsed(defaultTimeUnit), defaultTimeUnit);
+    }
+  }
+
+  /**
+   * Converts this {@code TimeUnit} to the equivalent {@code ChronoUnit}.
+   *
+   * @return the converted equivalent ChronoUnit
+   */
+  private static ChronoUnit toChronoUnit(TimeUnit timeUnit) {
+    switch (timeUnit) {
```

Comment:
It seems like there should be a better way to do this with the native libraries, but I see that wasn't introduced until Java 9.

Reply:
yeah this is only available in JDK9+, so I had to write up something for this